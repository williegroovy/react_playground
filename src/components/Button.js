import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const SButton = styled.button`
  font-family: quicksand;
  font-size: ${({ block }) => block ? 20 : 16 }px;
  width: 100px;
  font-weight: 500;
  background: #6bada7;
  border-radius: 4px;
  border: 0;
  color: ${({ block }) => block ? 'red' : 'white'};
  cursor: pointer;
  padding: 5px 15px 5px 15px;
  
  &:hover {
    ${({ animated }) => animated && 'animation: popAnimation 0.8s ease'};
  };
  
  &:focus {
    outline: none;
  }
  
  ${({ disabled }) => disabled && `
    background: #ccc;
    color: white;
    cursor: not-allowed;
  `};
  
  @keyframes popAnimation {
    50% { transform: scale(1.2)  }
  }
`;

const Button = ({ children, type='button', onClick, disabled, animated }) => (
    <SButton
      type={type}
      onClick={() => onClick && onClick()}
      animated={animated && !disabled}
      disabled={disabled}
    >
      {children}
    </SButton>
  );

Button.defaultProps = {
  type: 'button',
  onClick: false,
  disabled: false,
  animated: false
};

Button.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.string
  ]).isRequired,
  type: PropTypes.oneOf(['button', 'rest', 'submit']),
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  animated: PropTypes.bool
};

export default Button;