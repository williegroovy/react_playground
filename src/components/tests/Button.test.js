import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import Button from '../Button';

describe('<Button />', () => {
  test('should be defined', () => {
    expect(Button).toBeDefined();
  });

  test('should render correctly ' , () => {
    const wrapper = shallow(
      <Button>Hey, listen!!!</Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render button text' , () => {
    const wrapper = shallow(<Button>Click Me</Button>);
    expect(wrapper.children().text()).toBe('Click Me');
  });

  test('should be disabled when disabled prop is true' , () => {
    const wrapper = shallow(<Button disabled>Click Me</Button>);
    expect(wrapper.children().text()).toBe('Click Me');
  });

  test('should be animated if animated prop is true' , () => {
    const wrapper = shallow(<Button animated>Click Me</Button>);
    expect(wrapper.props().animated).toBeTruthy();
  });

  test('should be not be animated if disabled' , () => {
    const wrapper = shallow(<Button disabled animated>Click Me</Button>);
    expect(wrapper.props().animated).toBeFalsy();
  });

  test('should call onClick when button is clicked', () => {
    const clickHandler = jest.fn();
    const wrapper = shallow(<Button onClick={clickHandler}>Click Me</Button>);
    wrapper.simulate('click');
    expect(clickHandler.mock.calls.length).toBe(1);
  })
});