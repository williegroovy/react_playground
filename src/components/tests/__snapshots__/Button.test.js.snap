// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render correctly  1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    animated={false}
    disabled={false}
    onClick={null}
    type="button"
  >
    Hey, listen!!!
  </Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "animated": false,
      "buttonType": "primary",
      "children": "Hey, listen!!!",
      "disabled": false,
      "hidden": false,
      "onClick": [Function],
      "type": "button",
    },
    "ref": null,
    "rendered": "Hey, listen!!!",
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": false,
        "rules": Array [
          "
  width: 120px;
  background-color: ",
          [Function],
          ";
  font-family: quicksand;
  font-size: ",
          [Function],
          "px;
  font-weight: 500;
  text-align: center;
  border-radius: 4px;
  border: ",
          [Function],
          ";
  color: ",
          [Function],
          ";
  cursor: pointer;
  padding: 5px 15px 5px 15px;
  
  &:hover {
    ",
          [Function],
          ";
  };
  
  &:focus {
    outline: none;
  }
  
  ",
          [Function],
          ";
  
  @keyframes popAnimation {
    50% { transform: scale(1.2)  }
  }
  
  ",
          [Function],
          ";
",
        ],
      },
      "displayName": "styled.button",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "button",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "animated": false,
        "buttonType": "primary",
        "children": "Hey, listen!!!",
        "disabled": false,
        "hidden": false,
        "onClick": [Function],
        "type": "button",
      },
      "ref": null,
      "rendered": "Hey, listen!!!",
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": false,
          "rules": Array [
            "
  width: 120px;
  background-color: ",
            [Function],
            ";
  font-family: quicksand;
  font-size: ",
            [Function],
            "px;
  font-weight: 500;
  text-align: center;
  border-radius: 4px;
  border: ",
            [Function],
            ";
  color: ",
            [Function],
            ";
  cursor: pointer;
  padding: 5px 15px 5px 15px;
  
  &:hover {
    ",
            [Function],
            ";
  };
  
  &:focus {
    outline: none;
  }
  
  ",
            [Function],
            ";
  
  @keyframes popAnimation {
    50% { transform: scale(1.2)  }
  }
  
  ",
            [Function],
            ";
",
          ],
        },
        "displayName": "styled.button",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "button",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
